library/ssl_tls.c

```diff
@@ -8243,6 +8257,10 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
 
 exit:
 
+    if( check_cert_initiated && ret == 0 )
```

Comment:
@AndrzejKurek ,  what exactly we try to check in this line ?

Reply:
There are many cases in which the certificate is not checked. We're checking here that we hit a case where the "cert check" label has been executed, so a cert should be checked, and we're checking that ret is equal to zero (from any kind of source, like the previous `ret = ssl_remember_peer_pubkey( ssl, pk_start, pk_len );`). Only if it's zero, we check the additional flag - `ret_fi`.

Reply:
1. I don't see ret_fi in the code.
2. check_cert_initiated will be 1 cert_check label was execute, so why we are expecting 0 here ?

Reply:
1. I mean `ret_verify` - the dedicated variable for storing the result of `ssl_parse_certificate_verify`.
2. `crt_verify` is not called in all flows of this function. If it was called then the flow is as follows:
- first, check if `ret == 0`. This is to make sure that there are no errors from the normal flow of the function (like from `ssl_remember_peer_pubkey` or `ssl_remember_peer_crt_digest`).
- If there are no errors from the normal flow of the function - assign `ret_verify` to `ret` - to check if there was no verification error.
Then the normal flow of the function continues.
The reason for such solution was that in many situations the certificate is not verified, and checking if it should be verified or not at the end would not be so easy. Flow control also wouldn't be easy, as there are multiple different flows here. This is why `check_cert_initiated` was introduced.